### DISCUSSION TITLE
using Microsoft.Data.Sqlclient in a CmdLet, gives PlatformNotSupportedException

### DISCUSSION THREAD

**Discussion by bzuidgeest** [PowerShell/PowerShell#15360] [Q&A]
I trying to create a custom Cmdlet in c# for powershell 7. The CmdLet and one of the dependencies (a .net standard 2.0 lib) Both use Microsoft.Data.SqlClient. As soon as a try to use the CmdLet I get an error as soon as I try to open a sql connection:

customerDatabase = new SqlConnection(configuration.CustomerDatabase.ConnectionString);
customerDatabase.Open();

System.Exception
  HResult=0x80131500
  Message=Could not connect to customer Database

Inner Exception 1:
**PlatformNotSupportedException: Strings.PlatformNotSupported_DataSqlClient**

From what I can gather this seems to have something to do with .net not choosing the correct dll for the platform and falling back to some kind of dummy dll. But I cannot find much on how to fix it (that I am able to understand). Lot's of stuff about when it happens to console or webapps, but little to nothing on powershell and custom cmdlets, I have not been able to adapt a solution.

cmdlet dll is loaded like C:\Program Files\PowerShell\7\pwsh.exe -NoExit -Command "& Import-Module .\VaultTech.Powershell.dll -Verbose; Test-SqlCmdlet"

Anyone here that can give me the proper pointers?

**Comment by vexx32** [comment]
You'll most likely need to manually include and/or load the Microsoft.Data.SqlClient.dll before loading your cmdlet dll so that it's already present in the session before your cmdlet attempts to use those functions.

For cases like this, I usually recommend packaging the module as a PSM1 which then loads the DLLs explicitly in the order they need to be loaded.

**Reply by bzuidgeest** [reply]
I added an psd1 file to "preload" the dll

```
#
# Module manifest for module 'VaultTech.Powershell'
#
# Generated by: 
#
# Generated on: 7-5-2021
#

@{

# Script module or binary module file associated with this manifest.
RootModule = '.\VaultTech.PowerShell.dll'

# Version number of this module.
ModuleVersion = '0.0.1'

# Supported PSEditions
# CompatiblePSEditions = @()

# ID used to uniquely identify this module
GUID = '86547e5b-417a-452c-b7d5-a8ee92af4c39'

# Author of this module
Author = 'bart'

# Company or vendor of this module
CompanyName = 'Unknown'

# Copyright statement for this module
Copyright = '(c) bart. All rights reserved.'

# Assemblies that must be loaded prior to importing this module
RequiredAssemblies = @(
    "Microsoft.Data.SqlClient.dll"
)

# Functions to export from this module, for best performance, do not use wildcards and do not delete the entry, use an empty array if there are no functions to export.
FunctionsToExport = '*'

# Cmdlets to export from this module, for best performance, do not use wildcards and do not delete the entry, use an empty array if there are no cmdlets to export.
CmdletsToExport = '*'

# Variables to export from this module
VariablesToExport = '*'

# Aliases to export from this module, for best performance, do not use wildcards and do not delete the entry, use an empty array if there are no aliases to export.
AliasesToExport = '*'

# Private data to pass to the module specified in RootModule/ModuleToProcess. This may also contain a PSData hashtable with additional module metadata used by PowerShell.
PrivateData = @{

    PSData = @{

        # Tags applied to this module. These help with module discovery in online galleries.
        # Tags = @()

        # A URL to the license for this module.
        # LicenseUri = ''

        # A URL to the main website for this project.
        # ProjectUri = ''

        # A URL to an icon representing this module.
        # IconUri = ''

        # ReleaseNotes of this module
        # ReleaseNotes = ''

        # Prerelease string of this module
        # Prerelease = ''

        # Flag to indicate whether the module requires explicit user acceptance for install/update/save
        # RequireLicenseAcceptance = $false

        # External dependent modules of this module
        # ExternalModuleDependencies = @()

    } # End of PSData hashtable

} # End of PrivateData hashtable
}
```
Unfortunately that changed nothing. Except that it now automatically loads the dependcy dll into powershell. Error stays the same.

**Reply by bzuidgeest** [reply]
Ok it seems that I have to do many many add-type commands

copy the Microsoft.SqlClient.SNI.dll for the proper platform (x64 or x86) to the output folder
Add-Type -Path C:\Users\bzs\source\repos\Edu-IGA_IDvault\VaultTech\packages\Microsoft.Data.SqlClient.2.1.0\runtimes\win\lib\netcoreapp2.1\Microsoft.Data.SqlClient.dll;
Add-Type -Path C:\Users\bzs\.nuget\packages\microsoft.identity.client\4.21.1\lib\netcoreapp2.1\Microsoft.Identity.Client.dll;

And that get's sql working.

After that it keeps going on about wcf dll's

Seems that I have to do everything manually. I think I will just drop powershell as completely broken. And return to just making commandline tools.

**Comment by SeeminglyScience** [comment]
You probably need to compile for a more specific version.  For some packages, the `netstandard` version they distribute is filled with `=> throw PlatformNotSupportedException()` because they rely on an API that is not in `netstandard`.

**Comment by elthombre** [comment]
I need Microsoft.Data.SqlClient for a current automation against an Azure SQL database. I am having issues getting the package installed.

What is the status of this?

**Reply by rhubarb-geek-nz** [reply]
> What is the status of this?

I would recommend testing [rhubarb-geek-nz.SqlConnection](https://www.powershellgallery.com/packages/rhubarb-geek-nz.SqlConnection/2.0.2), this loads the official Microsoft assembly into its own ALC and works with PowerShell 7.4 and 7.5.
There is a [github project](https://github.com/rhubarb-geek-nz/SqlConnection) if you want to build it yourself.
